pkg/symbol/symbolsearcher/symbol_searcher.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package symbolsearcher
+
+import (
+	"debug/elf"
+	"errors"
+	"sort"
+	"strings"
+)
+
+type Searcher struct {
+	symbols []elf.Symbol
+}
+
+func New(syms []elf.Symbol) Searcher {
+	newSyms := make([]elf.Symbol, 0, len(syms))
+	for _, s := range syms {
+		if isFunction(s) {
+			newSyms = append(newSyms, s)
+		}
+	}
+
+	// slice stable sort to keep output consistent
+	sort.SliceStable(newSyms, func(i, j int) bool {
+		if newSyms[i].Value != newSyms[j].Value {
+			return newSyms[i].Value < newSyms[j].Value
+		}
+		return chooseBestSymbol(newSyms[i], newSyms[j])
+	})
+	return Searcher{
+		symbols: newSyms,
+	}
+}
+
+func (s Searcher) Search(addr uint64) (string, error) {
+	i := sort.Search(len(s.symbols), func(i int) bool {
+		sym := s.symbols[i]
+		return sym.Value > addr
+	})
+	if i == 0 ||
+		// addr < sym[i-1]
+		addr < s.symbols[i-1].Value {
+		return "", errors.New("failed to find symbol for address")
+	}
+
+	// sym[i-1] <= addr < sym[i]
+	i--
+	return s.symbols[i].Name, nil
+}
+
+func (s Searcher) PCRange() ([2]uint64, error) {
+	if len(s.symbols) == 0 {
+		return [2]uint64{}, errors.New("no symbols found")
+	}
+
+	return [2]uint64{
+		s.symbols[0].Value,
+		s.symbols[len(s.symbols)-1].Value + s.symbols[len(s.symbols)-1].Size,
+	}, nil
+}
+
+// copy from symbol-elf.c/elf_sym__is_function.
+func isFunction(s elf.Symbol) bool {
+	return elf.ST_TYPE(s.Info) == elf.STT_FUNC && s.Name != "" && s.Section != elf.SHN_UNDEF
+}
+
+// copy from symbol.c/choose_best_symbol.
+func chooseBestSymbol(syma, symb elf.Symbol) bool {
```

Comment:
⛏ This is just a nitpick; please feel free to ignore it.

I understand that this function directly translates to a C function. However, it would be nicer to change the style to be more Go idiomatic.